src/Core/src/AbstractApi.php

```diff
@@ -79,18 +79,27 @@ public function __construct($configuration = [], ?CredentialProvider $credential
         return $this->configuration;
     }
 
+    final public function presign(Input $input, ?\DateTimeInterface $expires = null): string
+    {
+        $request = $input->request();
+        $request->setEndpoint($this->getEndpoint($request->getUri(), $request->getQuery()));
+
+        $this->getSigner()->presign($request, $this->credentialProvider->getCredentials($this->configuration), $expires ? $expires : new \DateTimeImmutable('+60min'));
+
+        return $request->getEndpoint();
+    }
+
     abstract protected function getServiceCode(): string;
 
     abstract protected function getSignatureVersion(): string;
 
     abstract protected function getSignatureScopeName(): string;
 
-    final protected function getResponse(Request $request): Response
+    final protected function getResponse(Request $request, ?string $operation = null): Response
```

Comment:
in fact, this is not a `or`, but an extra header required by some S3 operation.
The information is not available in metadata file, and can't not be generated :-(

Nevertheless, I like the idea of object to avoid futures BC break.

Reply:
added a ContextRequest object